### PR TITLE
Pytest framework implementation with Task Runner GitHub Workflow

### PR DESCRIPTION
**Pull request by noopurintel**
This PR aims to add the `pytest` framework based test cases in OpenFL repository.

Changes done:

Added **end_to_end** folder inside **openfl/tests**. Please go through `README.md` file for complete structure and related details.
Added `task_runner_e2e.yml` workflow file inside **.github/workflows** - it will run various models for Task Runner API approach.
For now, the workflow is expected to run every night with below matrix combination:

1. model names - [torch_cnn_mnist, keras_cnn_mnist] # Other models have bugs associated with them
2. python version - [3.8, 3.9, 3.10]

Successful test run (11th Nov 07:20 pm IST) - https://github.com/securefederatedai/openfl/actions/runs/11779770937

Final (12th Nov 04:35 pm IST) - https://github.com/securefederatedai/openfl/actions/runs/11795815852

### CODE REVIEW

**Review by MasterSkepticista — COMMENTED** [review]
Suggest naming it as `tests/end_to_end`, and instances of `e2e` as `end_to_end`.

**Comment by noopurintel** [comment]
@MasterSkepticista - this name `openfl_e2e` was decided during one of the discussions. 
But, the name `end_to_end` would be easier to comprehend as you have suggested, let me check and get back on this.

File: test-requirements.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
 pytest==8.3.3
 pytest-asyncio==0.24.0
 pytest-mock==3.14.0
+ruamel.yaml
```
Comment (theakshaypant) [inline test-requirements.txt]: Nit: Do we not want to pin this to a specific version?

File: tests/end_to_end/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# Project Title
+
+This project is a machine learning workspace that includes various models and test suites. It is structured to facilitate the development, testing, and deployment of machine learning models.
+
+## Project Structure
```
Comment (theakshaypant) [inline tests/end_to_end/README.md]: General question regarding the test directory structure, would having the "utils" subdirectory outside directly under the "tests" directory not make more sense as not all of them seem very specific to e2e tests?
I might be missing the bigger picture with the planned testing structure hence the question.

**Review by theakshaypant — APPROVED** [review]
LGTM. 
Just a few general questions. 

Also how do we feel about having the model accuracy to be logged along with test success (or failure) in the workflow? (Can be easily picked up from the aggregator logs.)

File: tests/end_to_end/README.md
Comment (MasterSkepticista) [inline tests/end_to_end/README.md:1]: [Improperly](https://github.com/securefederatedai/openfl/blob/eda91a5f7bbbb7fcf74549656e5123c9ead617b0/tests/end_to_end/README.md) formatted.

File: tests/end_to_end/__init__.py
Comment (MasterSkepticista) [inline tests/end_to_end/__init__.py:1]: Use consistent copyright headers across repository (applies to all impacted files in this PR). Example [here](https://github.com/securefederatedai/openfl/blob/e87627f14cdeae417ace9547b4d085b331973292/openfl/component/__init__.py#L1C1-L2C38).

File: test-requirements.txt
Comment (MasterSkepticista) [inline test-requirements.txt:1]: Any particular reason to use `ruamel`? I believe PyYAML is default in OpenFL.

File: tests/end_to_end/conftest.py
Comment (MasterSkepticista) [inline tests/end_to_end/conftest.py:1]: All impacted modules: Provide a [module level docstring](https://google.github.io/styleguide/pyguide.html#3821-test-modules) that conveys purpose and objectives of the test/module.

File: tests/end_to_end/pytest.ini [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+testpaths = test_suites
+junit_family = xunit2
+results_dir = results
+log_level = INFO
+markers =
+    torch_cnn_mnist: mark a test as a torch CNN MNIST test.
+    keras_cnn_mnist: mark a test as a Keras CNN MNIST test.
+    torch_cnn_histology: mark a test as a torch CNN histology test.
+    tf_2dunet: mark a test as a tf 2D U-Net test.
+    tf_cnn_histology: mark a test as a tf CNN histology test.
```
Comment (MasterSkepticista) [inline tests/end_to_end/pytest.ini]: Request removing TF tests. TF tutorials will also be dispositioned soon from OpenFL. Going forward, it will be Keras (with all backends), PyTorch and Jax.

**Review by MasterSkepticista — CHANGES_REQUESTED** [review]
I have not reviewed the actual tests yet.

File: test-requirements.txt (hunk 1 above)
Comment (teoparvanov) [reply test-requirements.txt]: I agree with @theakshaypant on pinning a specific version, otherwise we may end up with strange intermittent issues that would waste cycles to debug.

File: tests/end_to_end/README.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
+# Project Title
+
+This project is a machine learning workspace that includes various models and test suites. It is structured to facilitate the development, testing, and deployment of machine learning models.
```
Comment (teoparvanov) [inline tests/end_to_end/README.md]: Isn't it rather aimed at integration testing of OpenFL?

File: tests/end_to_end/models/participants.py [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+
+        with open(self.plan_path, "w+") as write_file:
+            yaml.dump(data, write_file)
+
+        log.info(
+            f"Modified the plan to train the model for collaborators {self.num_collaborators} and {self.rounds_to_train} rounds"
+        )
+        return True
+
+    def initialize_plan(self, agg_domain_name):
```
Comment (teoparvanov) [inline tests/end_to_end/models/participants.py:145]: I suggest setting `"localhost"` as the default `agg_domain_name`. This is suitable for testing on a single VM, and may also save you some of the Intel proxy headaches 😉 . Later on, you can pass an actual FQDN in case the same tests would support a distributed deployment.

File: tests/end_to_end/utils/federation_helper.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
+# Copyright 2024-2025 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import time
+import concurrent.futures
+
+from tests.end_to_end.utils.logger import logger as log
+
+
+def perform_csr_operations(fed_obj):
```
Comment (teoparvanov) [inline tests/end_to_end/utils/federation_helper.py]: I'd simply call this `setup_pki()`

File: tests/end_to_end/utils/federation_helper.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    """
+    Perform CSR operations like generating sign request, certifying request, etc.
```
Comment (teoparvanov) [inline tests/end_to_end/utils/federation_helper.py]: => Setup PKI for trusted communication within the federation

File: tests/end_to_end/models/participants.py [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+                log.error(f"Failed to generate the sign request: {error}")
+                raise Exception(f"Failed to generate the sign request: {error}")
+
+            log.info(f"Generated a sign request for {self.name}")
+        except Exception as e:
+            log.error(f"Failed to generate the sign request: {e}")
+            raise e
+        return True
+
+    def certify_request(self):
```
Comment (teoparvanov) [inline tests/end_to_end/models/participants.py]: I suggest moving the `certify_request` method to `ModelOwner`, as he/she is effectively acting as the certificate authority and root of trust (certifying the aggregator, as well as the collaborator nodes).

File: tests/end_to_end/models/participants.py [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+                log.error(f"Failed to certify the aggregator request: {error}")
+                raise Exception(f"Failed to certify the aggregator request: {error}")
+
+            log.info(f"CA signed the request from {self.name}")
+        except Exception as e:
+            log.error(f"Failed to certify the aggregator request : {e}")
+            raise e
+        return True
+
+    def sign_collaborator_csr(self, collaborator_name):
```
Comment (teoparvanov) [inline tests/end_to_end/models/participants.py]: Same here, let's move `sign_collaborator_csr` to `ModelOwner` as the CA representative.

**Review by teoparvanov — APPROVED** [review]
This is awesome, thank you @noopurintel ! Very nice abstractions around `ModelOwner`, `Collaborator` and `Aggregator`, which contribute to a great experience both reading and writing the actual tests.

File: tests/end_to_end/models/participants.py (hunk 9 above)
Comment (noopurintel) [reply tests/end_to_end/models/participants.py]: This is done.

File: tests/end_to_end/models/participants.py (hunk 8 above)
Comment (noopurintel) [reply tests/end_to_end/models/participants.py]: This is done.

File: tests/end_to_end/utils/federation_helper.py (hunk 7 above)
Comment (noopurintel) [reply tests/end_to_end/utils/federation_helper.py]: Corrected the doc-string, also modified at other relevant places.

File: tests/end_to_end/utils/federation_helper.py (hunk 6 above)
Comment (noopurintel) [reply tests/end_to_end/utils/federation_helper.py]: Done as suggested.

File: tests/end_to_end/README.md (hunk 4 above)
Comment (noopurintel) [reply tests/end_to_end/README.md]: Modified the `README.md` file with proper title and description.

File: tests/end_to_end/models/participants.py (hunk 5 above)
Comment (noopurintel) [reply tests/end_to_end/models/participants.py:145]: The intention is to use an actual FQDN for various scenarios, thus keeping the value as `aggregator` and adding `hosts` file entry.

File: tests/end_to_end/pytest.ini (hunk 3 above)
Comment (noopurintel) [reply tests/end_to_end/pytest.ini]: Removed tf based tests.

File: tests/end_to_end/__init__.py
Comment (noopurintel) [reply tests/end_to_end/__init__.py:1]: This is fixed.

File: tests/end_to_end/README.md
Comment (noopurintel) [reply tests/end_to_end/README.md:1]: This is fixed.

File: tests/end_to_end/README.md (hunk 2 above)
Comment (noopurintel) [reply tests/end_to_end/README.md]: We discussed about it. As the files are right now utilised in the end_to_end test module, thus keeping them within the folder. In near future, based on the expansion, we will look for moving out these files, if required.

File: test-requirements.txt
Comment (noopurintel) [reply test-requirements.txt:1]: Replaced it with pyyaml and removed the entry from test-requirements.txt

File: tests/end_to_end/conftest.py
Comment (noopurintel) [reply tests/end_to_end/conftest.py:1]: Added for every participant - model owner, aggregator and collaborator in `participants.py`

File: test-requirements.txt (hunk 1 above)
Comment (noopurintel) [reply test-requirements.txt]: I have replaced it with pyyaml which is already present for OpenFL. Removed this entry.

**Comment by noopurintel** [comment]
@theakshaypant  - Fyi, we are planning to add accuracy and other details after every test run in subsequent PRs.

File: tests/end_to_end/models/participants.py [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+        log.info(f"Workspace path: {workspace_path}")
+        if os.path.exists(workspace_path):
+            self.workspace_path = workspace_path
+            log.info(f"Workspace path: {self.workspace_path}")
+        else:
+            log.error(f"Workspace {workspace_name} does not exist in {results_dir}")
+            raise FileNotFoundError(f"Workspace {workspace_name} does not exist in {results_dir}")
+        return self.workspace_path
+
+    def setup_pki(self, collaborator_name):
```
Comment (teoparvanov) [inline tests/end_to_end/models/participants.py]: My apologies for the confusion, I back-track on my suggestion for renaming, as this method only signs the collaborator's CSR (rather than setting up the entire PKI). Let's rename to `certify_collaborator()`, for symmetry with `certify_aggregator()` suggested below.

File: tests/end_to_end/models/participants.py [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+                log.error(f"Failed to certify the workspace: {error}")
+                raise Exception(f"Failed to certify the workspace: {error}")
+
+            log.info(f"Certified the workspace {self.workspace_name}")
+        except Exception as e:
+            log.error(f"Failed to certify the workspace: {e}")
+            raise e
+        return True
+
+    def certify_agg_request(self, agg_domain_name):
```
Comment (teoparvanov) [inline tests/end_to_end/models/participants.py]: For the sake of symmetry, let's rename this to simply `certify_aggregator()`.

**Review by teoparvanov — COMMENTED** [review]
A final comment on method naming...

File: tests/end_to_end/models/participants.py (hunk 10 above)
Comment (noopurintel) [reply tests/end_to_end/models/participants.py]: Done. Renamed to `certify_collaborator` and `certify_aggregator`

File: tests/end_to_end/models/participants.py (hunk 11 above)
Comment (noopurintel) [reply tests/end_to_end/models/participants.py]: This is done

**Review by teoparvanov — APPROVED** [review]

File: .github/workflows/task_runner_e2e.yml [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+  test_run:
+    name: test
+    runs-on: ubuntu-22.04
+    timeout-minutes: 120 # 2 hours
+    strategy:
+      matrix:
+        # There are open issues for some of the models, so excluding them for now:
+        # model_name: [ "torch_cnn_mnist", "keras_cnn_mnist", "torch_cnn_histology" ]
+        model_name: [ "torch_cnn_mnist", "keras_cnn_mnist" ]
+        python_version: [ "3.8", "3.9", "3.10" ]
```
Comment (MasterSkepticista) [inline .github/workflows/task_runner_e2e.yml:41]: Python version test in an end-to-end workflow seems unnecessary/orthogonal. Ablation across models, backends, plan values seem appropriate - as done in the rest of the PR.

@teoparvanov WDYT?

File: tests/end_to_end/README.md [hunk 13]
```diff
@@ -0,0 +1,62 @@
+# End-to-end Pytest Framework
```
Comment (MasterSkepticista) [inline tests/end_to_end/README.md:1]: Nit: "End-to-End testing of Taskrunner API using PyTest"

File: .github/workflows/task_runner_e2e.yml [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+      run: |
+        python -m pip install --upgrade pip
+        pip install .
+        pip install -r test-requirements.txt
+
+    - name: Add runner IP to /etc/hosts
+      id: add_runner_ip
+      run: |
+        sudo echo "127.0.0.1 aggregator" | sudo tee -a /etc/hosts
+        echo "Added runner IP to /etc/hosts"
```
Comment (MasterSkepticista) [inline .github/workflows/task_runner_e2e.yml]: Similar to Teo's previous comment - can we stick to `localhost` as the FQDN? Reduces complexity.

File: tests/end_to_end/README.md [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+```
+results
+    ├── <workspace_name>    # Based on the workspace name provided during test run.
+    ├── results.xml         # Output file in JUNIT.
+    └── deployment.log      # Log file containing step by step test progress.
+```
+
+## Contribution
+Please ensure that you have tested your changes thoroughly before submitting a pull request.
```
Comment (MasterSkepticista) [inline tests/end_to_end/README.md]: Point this to [CONTRIBUTING.md](https://github.com/securefederatedai/openfl/blob/develop/CONTRIBUTING.md)

File: tests/end_to_end/README.md [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+## License
+This project is licensed under [Apache License Version 2.0](LICENSE). By contributing to the project, you agree to the license and copyright terms therein and release your contribution under these terms.
```
Comment (MasterSkepticista) [inline tests/end_to_end/README.md]: Not necessary.

File: tests/end_to_end/utils/logger.py
Comment (MasterSkepticista) [inline tests/end_to_end/utils/logger.py:1]: Could re-use OpenFL logging handler?

**Review by MasterSkepticista — CHANGES_REQUESTED** [review]
Minor changes.

File: tests/end_to_end/README.md (hunk 13 above)
Comment (payalcha) [reply tests/end_to_end/README.md:1]: This is very specific... we are planning to accommodate Workflow APIs as well in near future.

Comment (noopurintel) [reply tests/end_to_end/README.md:1]: Correct, this is a generic title meant to cover various APIs using pytest framework.

File: tests/end_to_end/utils/logger.py
Comment (noopurintel) [reply tests/end_to_end/utils/logger.py:1]: We are using `logging` module itself just like existing OpenFL modules, we customised it to print other info as well.
For example - date, log level, file name, function name etc.

```
2024-11-12 06:22:19 - INFO: [participants.py - start]: Starting collaborator1

2024-11-12 06:22:19 - INFO: [participants.py - start]: Starting collaborator2

2024-11-12 06:22:19 - INFO: [participants.py - start]: Starting aggregator
```

File: .github/workflows/task_runner_e2e.yml (hunk 14 above)
Comment (noopurintel) [reply .github/workflows/task_runner_e2e.yml]: Done. Using `localhost` now. 
Successful test run here - https://github.com/securefederatedai/openfl/actions/runs/11792318841

File: tests/end_to_end/README.md (hunk 16 above)
Comment (noopurintel) [reply tests/end_to_end/README.md]: Done.

File: tests/end_to_end/README.md (hunk 15 above)
Comment (noopurintel) [reply tests/end_to_end/README.md]: Done.

File: .github/workflows/task_runner_e2e.yml (hunk 12 above)
Comment (teoparvanov) [reply .github/workflows/task_runner_e2e.yml:41]: Hi @MasterSkepticista , in the PR pipeline we do run the taskrunner tests (end-to-end) against different Python versions:
https://github.com/securefederatedai/openfl/blob/develop/.github/workflows/taskrunner.yml#L24

Also, the workspaces covered here would usually be the entry point for newcomers, so we need to make sure they are compatible with a (relatively) wide range of python environments.

In conclusion, I'm ok with the approach suggested by @noopurintel here. We can always adjust down the road - esp. if the job's execution time becomes excessive as we add more workspaces to be validated.

Comment (noopurintel) [reply .github/workflows/task_runner_e2e.yml:41]: Marking it resolved as no change is required at the moment.

File: tests/end_to_end/utils/logger.py
Comment (MasterSkepticista) [inline tests/end_to_end/utils/logger.py:1]: OpenFL logging [includes](https://github.com/securefederatedai/openfl/blob/3301e577856d6c889bb33c405e7e2f69c0c2e5e1/openfl/interface/cli.py#L34) fileno/function/module names. Intent is to reuse functionality.
Alternatively, if there are pieces that could go into core OpenFL, please consider that too.

File: tests/end_to_end/utils/federation_helper.py [hunk 17]
```diff
@@ -0,0 +1,132 @@
+# Copyright 2020-2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import time
+import concurrent.futures
+
+from tests.end_to_end.utils.constants import SUCCESS_MARKER
+from tests.end_to_end.utils.logger import logger as log
```
Comment (MasterSkepticista) [inline tests/end_to_end/utils/federation_helper.py]: Can we not instantiate a new logger?

```python
import logging

log = logging.getLogger(__name__)
```

It will reuse the defined handler, AFAIK.

**Review by MasterSkepticista — APPROVED** [review]
Approving with reservations - I am personally not confident about the scope of what this test framework will achieve for "Taskrunner" interface, and how it helps beyond existing tests. I may be wrong about this, so I agree with merging this PR.

I do see the value of this framework for Workflow API though, to be clear.

Thanks for the additions @noopurintel and @payalcha.

File: tests/end_to_end/utils/logger.py
Comment (noopurintel) [reply tests/end_to_end/utils/logger.py:1]: Yes - this function `setup_logging` can be reused, just that its position is specific here. 
Is it possible to have this directly inside `openfl/utilities` instead of `openfl/openfl/interface` - to avoid dependency on specific module for a generic functionality ?
We can discuss on this and arrive at a proper solution.

File: tests/end_to_end/utils/federation_helper.py (hunk 17 above)
Comment (noopurintel) [reply tests/end_to_end/utils/federation_helper.py]: Yes, this works and is comparatively better. I have modified the files to include this change.
